Add ZLIB support

ekopylova opened the pull request:
This major update includes:
- [KSEQ](http://lh3lh3.users.sourceforge.net/parsefastq.shtml) library, allow compressed reads input (gzip) [resolves issue #94 and #71]
- support both mmap and generic buffer input
- clean up compiler warnings

wasade commented on .travis.yml:
```diff
@@ -14,11 +14,15 @@ before_install:
   - conda update --yes conda
 install:
-  - conda create --yes -n env_name python=2.7 numpy scipy
+  - conda create --yes -n env_name python=2.7 numpy scipy zlib
```
just curious, why py2?

ekopylova replied:
Probably can be py3 as well, haven't tested .. need python for the tests ```test_sortmerna.py``` and ```test_sortmerna_zlib.py```

wasade commented on ChangeLog:
```diff
@@ -1,3 +1,9 @@
+    - update ALP library to version 1.95
+    - add [KSEQ](http://lh3lh3.users.sourceforge.net/parsefastq.shtml) library, allow compressed reads input
+    - support both mmap and generic buffer input
+    - remove type warnings
```
compilation warnings?

ekopylova replied:
Yes, will update.

wasade commented on include/kseq_load.hpp:
```diff
@@ -0,0 +1,64 @@
+
+
+/*! @fn load_reads()
+    @brief load reads into buffer using kseq library
```
is the buffer being modified in place, or is the `char**` that's returned the loaded buffer?

ekopylova replied:
In this function ```char** load_reads()``` a 'buffer' (or 1D array of size `full_file_size`) ```char* raw = new char[full_file_size]()``` is created to store all reads, it is modified only in this function. An additional 2D array ```char** reads = new char*[number_total_read*2]()``` is created (and returned) to store pointers to each label and sequence in the ```raw``` buffer.

After reads are loaded into ```char* raw``` and ```char** reads``` pointers are setup pointing to sequences in ```raw```, sequence information can be extracted by accessing ```reads``` directly, ex. ```reads[0]``` returns ```>seq1``` and ```reads[1]``` returns ```ACTGTA..```

ekopylova replied:
I just modified the input from ```char* raw``` to ```char*& raw``` since I'm freeing the memory outside the function. Also, in case we need to modify the pointer later on. Thanks for `pointing` my attention to this :)

ekopylova commented on src/kseq_load.cpp:
```diff
@@ -0,0 +1,135 @@
+    exit(EXIT_FAILURE);
+  }
+#ifdef HAVE_LIBZ
+  // Count total number of reads and their combined length
```
Comment should be removed (same for one below). This computation is performed earlier now in ```compute_read_stats()``` from paralleltrfaversal.cpp

wasade commented on src/kseq_load.cpp:
```diff
@@ -0,0 +1,135 @@
+  if ( raw == NULL )
+  {
+    fprintf(stderr,"    %sERROR%s: [Line %d: %s] could not allocate memory for reference sequence buffer\n",
+                   "\033[0;31m","\033[0m", __LINE__, __FILE__);
```
always hated escape codes :/ do you think it would be useful to do them as `#define`'s?

ekopylova replied:
yes, good idea

wasade commented on src/kseq_load.cpp:
```diff
@@ -0,0 +1,135 @@
+ *******************************************************/
+char**
+load_reads(char* inputreads,
+           char* raw,
```
on input, is `raw` expected to be `NULL`? If so, would it make sense to assert that to avoid a potential memory leak?

wasade commented on src/kseq_load.cpp:
```diff
@@ -0,0 +1,135 @@
+    // sequence label
+    memcpy(input_ptr, seq->name.s, seq->name.l);
+    input_ptr += seq->name.l;
+    // comment
```
that is indeed a comment... ;)

wasade commented on src/kseq_load.cpp:
```diff
@@ -0,0 +1,135 @@
+    // comment
+    if (seq->comment.l)
+    {
+      *input_ptr++ = ' ';
```
Just to check, is this safe in the event that a sequence ID and comment are separated by multiple white spaces?

ekopylova replied:
This should be fine. We allocated ```full_file_size``` bytes (the size of the input reads file) to the buffer to which ```input_ptr``` is writing . Assuming only 1 space will always be an underestimate to ```full_file_size```. If there are 2 or more spaces in the original file, our buffer will just not be fully saturated. The end of input text is terminated by ```*input_ptr = '\0';```, so that's all we look for when reading from the buffer. Does that answer your question?

wasade commented on src/kseq_load.cpp:
```diff
@@ -0,0 +1,135 @@
+    *input_ptr++ = '\n';
+    // sequence
+    reads[i++] = input_ptr;
+    memcpy(input_ptr, seq->seq.s, seq->seq.l);
```
does this handle multiline fasta?

ekopylova replied:
yes, ```seq->seq.s``` is returned by the [KSEQ[(http://lh3lh3.users.sourceforge.net/parsefastq.shtml) library, which formats multiline to single, ```seq->seq.l``` gives the length of sequence without newlines.

wasade commented on src/kseq_load.cpp:
```diff
@@ -0,0 +1,135 @@
+  finalnt = input_ptr;
+  input_ptr++;
+  /* TEST */
+  //char* st = raw;
```
should theese comments be removed?

ekopylova replied:
yes :)

wasade commented on src/kseq_load.cpp:
```diff
@@ -0,0 +1,135 @@
+#endif
+  int l;
+  int i = 0;
+  char* input_ptr = raw;
```
it doesn't look like `raw` is used, would it make sense to just create `input_ptr` on like 50?

ekopylova replied:
hmm, ```raw``` is being passed to ```load_reads()``` as ```char*& raw```, meaning we will use the address to free the memory outside the function. Since ```input_str``` is being incremented, if we set ```input_str``` to ```raw```, this will cause an error in memory deallocation outside the function (since pointer will not longer point to the allocated address).

wasade commented on src/main.cpp:
```diff
@@ -164,11 +175,17 @@ void printlist()
   printf("                                        (maximum -m INT is %lu)\n",(((maxpages_gv/2)*pagesize_gv)/1048576));
+  printf("                                        (NOTE: If this option is chosen, reads will be loaded using\n");
+  printf("                                         mmap rather than the default generic stream buffer; if the\n");
+  printf("                                         input reads are in compressed format, this option will be\n");
```
can compressed data not use mmap?

ekopylova replied:
It may be possible with Boost C++ libraries but didn't want to add another external dependency. Also, since mmap maps pages into memory, at the moment we have a complex algorithm to parse sequences overlapping two pages. I would like to stay away from that for compressed files :|

wasade commented on src/main.cpp:
```diff
@@ -236,6 +253,8 @@ main(int argc,
   bool min_lis_gv_set = false;
   bool num_alignments_gv_set = false;
   bool best_gv_set = false;
+  bool have_reads = false;
```
not advocating change here, but ever seen Linus Torvalds rant on bool?

wasade commented on src/main.cpp:
```diff
@@ -1497,33 +1530,30 @@ main(int argc,
-                     exit_early);
-  
+  paralleltraversal(readsfile,
+                    have_reads_gz,
```
this arguments list is getting large, would it make sense to pass around a struct?

ekopylova replied:
The issue here is probably the function is doing too much and need to refactor. I plan to do this in the next PR when I work on the global alignment integration and substitution of alignment libraries.

wasade commented on src/mmap.cpp:
```diff
@@ -49,22 +49,22 @@ mmap_reads(off_t partial_file_size,
            int32_t &offset_pair_from_top,
```
since there was a change in type on line 52, should this also be uint? 

is mmap support limited to 2**32 files?

ekopylova replied:
mmap is limited by the virtual address space, so should be 2**32 bytes for a 32-bit system and 2**64 for a 64-bit

wasade commented on src/mmap.cpp:
```diff
@@ -111,27 +109,24 @@ mmap_reads(off_t partial_file_size,
         // count the number of strings in the file section
-        for ( int64_t i = reads_offset_f; i < partial_file_size-reads_offset_e-2; i++ ) if ( raw[i] == '>' ) strs++;
-
+        for ( int64_t i = reads_offset_f; i < partial_file_size-reads_offset_e-2; i++ ) if ( raw[i] == '>' )
```
uint?

wasade replied:
would it be possible to break the if-statement out onto a separate line? this is a bit complex to look at.

also, what is the significance of the `-2`?

wasade replied:
would `raw[i] == '>'` handle the following as expected?

```text
>foo>imaterriblefasta
atgatatata
>bar
gtgtgtgt
```

ekopylova replied:
The significance of ```-2``` is to remove the "\n>" characters from search at the end of the mapped file section. I've added this in the comment.

for the code:
```
for ( int64_t i = reads_offset_f; i < partial_file_size-reads_offset_e-2; i++ )
```

we have:
```
file section 1
>seq1\n
TGTGA\n
    ^ (2) the -2 reduces the search space to here
>seq2\n
^ (1) reads_offset_e sets the search space to here
AA ..
```

wasade commented on src/mmap.cpp:
```diff
@@ -384,37 +354,28 @@ mmap_reads(off_t partial_file_size,
         while ( *line++ != '\n' ) readlen++;
-        if ( readlen >= 20 ) readlen < minlenread ? minlenread = readlen : minlenread;
-                  
+        if ( readlen >= 20 ) readlen < minlenread ? minlenread = readlen : minlenread; 
```
20?

wasade replied:
the ternary is a bit difficult to follow...

ekopylova replied:
20 has been replaced by the seed length (LNWIN). Added an inline comment to describe the line.

wasade commented:
Possible to add a check for `aclocal` into `configure`? Getting the following on barnacle:

```bash
15:29:00 (dtmcdonald@barnacle):sortmerna@ekopylova-add_zlib$ make -j 8
CDPATH="${ZSH_VERSION+.}:" && cd . && /bin/sh /home/mcdonadt/smr2_testing/sortmerna/missing aclocal-1.15
/home/mcdonadt/smr2_testing/sortmerna/missing: line 81: aclocal-1.15: command not found
WARNING: 'aclocal-1.15' is missing on your system.
         You should only need it if you modified 'acinclude.m4' or
         'configure.ac' or m4 files included by 'configure.ac'.
         The 'aclocal' program is part of the GNU Automake package:
         <http://www.gnu.org/software/automake>
         It also requires GNU Autoconf, GNU m4 and Perl in order to run:
         <http://www.gnu.org/software/autoconf>
         <http://www.gnu.org/software/m4/>
         <http://www.perl.org/>
make: *** [aclocal.m4] Error 127
```

wasade commented:
Interesting. `build.sh` works though. However, `bash ./build.sh` is not mentioned in `INSTALL`

wasade commented:
...sorry for so many comments. `build.sh` proceeded further than `./configure && make`, however I then got:

```bash
(CDPATH="${ZSH_VERSION+.}:" && cd . && /bin/sh /home/mcdonadt/smr2_testing/sortmerna/missing autoheader)
aclocal.m4:17: warning: this file was generated for autoconf 2.69.
You have another version of autoconf.  It may work, but is not guaranteed to.
If you have problems, you may need to regenerate the build system entirely.
To do so, use the procedure documented by the package, typically 'autoreconf'.
configure.ac:6: error: Autoconf version 2.65 or higher is required
aclocal.m4:437: AM_INIT_AUTOMAKE is expanded from...
configure.ac:6: the top level
autom4te: /usr/bin/m4 failed with exit status: 63
autoheader: '/usr/bin/autom4te' failed with exit status: 63
WARNING: 'autoheader' is probably too old.
         You should only need it if you modified 'acconfig.h' or
         'configure.ac' or m4 files included by 'configure.ac'.
         The 'autoheader' program is part of the GNU Autoconf package:
         <http://www.gnu.org/software/autoconf/>
         It also requires GNU m4 and Perl in order to run:
         <http://www.gnu.org/software/m4/>
         <http://www.perl.org/>
make: *** [include/config.h.in] Error 63
```

ekopylova commented:
You're right, ```bash ./build.sh``` is not mentioned in the INSTALL, I will update that. It is in the README.md instructions however.

ekopylova commented:
So, if you start from the sortmerna directory, do these commands fail?

```
make clean
bash ./build.sh
```

ekopylova commented:
The ```build.sh``` is a workaround to the fact that autotools configure requires different timestamps for various files, whereas git gives the same times to all files on pull.

wasade commented:
From barnacle:

```bash
$ make clean
$ bash ./build.sh --without-zlib
...snip...
/usr/bin/ld: skipping incompatible /usr/lib/libm.so when searching for -lm
/usr/bin/ld: skipping incompatible /usr/lib/libm.so when searching for -lm
/usr/bin/ld: skipping incompatible /usr/lib/libpthread.so when searching for -lpthread
/usr/bin/ld: skipping incompatible /usr/lib/libpthread.so when searching for -lpthread
/usr/bin/ld: skipping incompatible /usr/lib/libc.so when searching for -lc
/usr/bin/ld: skipping incompatible /usr/lib/libc.so when searching for -lc
(CDPATH="${ZSH_VERSION+.}:" && cd . && /bin/sh /home/mcdonadt/smr2_testing/sortmerna/missing autoheader)
aclocal.m4:17: warning: this file was generated for autoconf 2.69.
You have another version of autoconf.  It may work, but is not guaranteed to.
If you have problems, you may need to regenerate the build system entirely.
To do so, use the procedure documented by the package, typically 'autoreconf'.
configure.ac:6: error: Autoconf version 2.65 or higher is required
aclocal.m4:437: AM_INIT_AUTOMAKE is expanded from...
configure.ac:6: the top level
autom4te: /usr/bin/m4 failed with exit status: 63
autoheader: '/usr/bin/autom4te' failed with exit status: 63
WARNING: 'autoheader' is probably too old.
         You should only need it if you modified 'acconfig.h' or
         'configure.ac' or m4 files included by 'configure.ac'.
         The 'autoheader' program is part of the GNU Autoconf package:
         <http://www.gnu.org/software/autoconf/>
         It also requires GNU m4 and Perl in order to run:
         <http://www.gnu.org/software/m4/>
         <http://www.perl.org/>
make: *** [include/config.h.in] Error 63
```

...same error with the other permutations. Also, I observed the following warnings:

```bash
src/indexdb.cpp: In function ‘int main(int, char**)’:
src/indexdb.cpp:1545: warning: format ‘%llu’ expects type ‘long long unsigned int’, but argument 3 has type ‘uint64_t’

src/load_index.cpp: In function ‘void load_index(char*, std::string, kmer*&, kmer_origin*&, uint32_t&, uint32_t)’:
src/load_index.cpp:369: warning: dereferencing type-punned pointer will break strict-aliasing rules
src/load_index.cpp:421: warning: dereferencing type-punned pointer will break strict-aliasing rules
src/load_index.cpp:369: warning: dereferencing pointer ‘dst.161’ does break strict-aliasing rules
src/load_index.cpp:369: note: initialized from here
src/load_index.cpp:421: warning: dereferencing pointer ‘dst.161’ does break strict-aliasing rules
src/load_index.cpp:421: note: initialized from here

src/paralleltraversal.cpp:264: warning: format ‘%lld’ expects type ‘long long int’, but argument 3 has type ‘uint64_t’
src/paralleltraversal.cpp:1303: warning: format ‘%llu’ expects type ‘long long unsigned int’, but argument 3 has type ‘uint64_t’
src/paralleltraversal.cpp:1735: warning: format ‘%llu’ expects type ‘long long unsigned int’, but argument 3 has type ‘uint64_t’
src/paralleltraversal.cpp:1866: warning: format ‘%llu’ expects type ‘long long unsigned int’, but argument 3 has type ‘uint64_t’
src/paralleltraversal.cpp:1869: warning: format ‘%llu’ expects type ‘long long unsigned int’, but argument 3 has type ‘uint64_t’
src/paralleltraversal.cpp:1872: warning: format ‘%llu’ expects type ‘long long unsigned int’, but argument 3 has type ‘uint64_t’
src/paralleltraversal.cpp:1873: warning: format ‘%llu’ expects type ‘long long unsigned int’, but argument 3 has type ‘uint64_t’
src/paralleltraversal.cpp:1885: warning: format ‘%llu’ expects type ‘long long unsigned int’, but argument 3 has type ‘uint64_t’
```

ekopylova commented:
I've removed the ```build.sh``` script and replaced it with ```autogen.sh```. With recent updates to autotools, it doesn't seem possible to compile the code cloning from github unless ```autoheader``` is installed. As the case with ```vsearch```, if users want to compile the code from cloned source they will need to run ```autogen.sh``` first, followed by ```configure; make```. We will provide binaries with each release on github and a tarball with distribution code on the bioinfo website (also github, if we can upload tarballs).
